auth/auth.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// TokenType represents one of the possible token Types
+type TokenType string
+
+// ClaimConverter converts claims to another domain object for saving into Context
+type ClaimConverter func(*ScopedClaims) interface{}
+
+// ScopedClaims struct to store the scoped claim related things
+type ScopedClaims struct {
+	jwt.StandardClaims
+	Scope string `json:"scope,omitempty"`
+	// Drone fields
+	Type TokenType `json:"type,omitempty"`
+	Text string    `json:"text,omitempty"`
+}
+
+type options struct {
+	tokenStore tokenStore
+	extractors []jwtRequest.Extractor
+}
+
+type Option interface {
+	apply(o *options)
+}
+
+type ContextSetter func(c context.Context, value interface{}) context.Context
+
+type authOptionFunc func(o *options)
+
+func (fn authOptionFunc) apply(o *options) {
+	fn(o)
+}
+
+func TokenStoreAuthOption(tokenStore tokenStore) Option {
+	return authOptionFunc(func(o *options) {
+		o.tokenStore = tokenStore
+	})
+}
+
+func ExtractorAuthOption(extractor jwtRequest.Extractor) Option {
+	return authOptionFunc(func(o *options) {
+		o.extractors = append(o.extractors, extractor)
+	})
+}
+
+// JWTAuth returns a new JWT authentication Gin Handler
+// Parameters:
+// - signingKey - the HMAC JWT token signing key
+// - claimConverter - converts the JWT token into a JWT claim object, which will be saved into the request context
+// - extractors (optional) - additional token extractors to use besides
+// 							github.com/dgrijalva/jwt-go/request.OAuth2Extractor
+// - tokenStore (optional) - checks if the incoming JWT Bearer token's ID is present in this TokenStore
+//							(can be nil, which pypasses the check)
+func JWTAuthHandler(
+	signingKey string,
+	claimConverter ClaimConverter,
+	contextSetter ContextSetter,
+	opts ...Option) gin.HandlerFunc {
+
+	var o options
+	for _, opt := range opts {
+		opt.apply(&o)
+	}
+
+	signingKeyBase32 := []byte(base32.StdEncoding.EncodeToString([]byte(signingKey)))
+
+	hmacKeyFunc := func(token *jwt.Token) (interface{}, error) {
+		// Don't forget to validate the alg is what you expect
+		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Method.Alg())
+		}
+		return signingKeyBase32, nil
+	}
+
+	extractor := append(jwtRequest.MultiExtractor{jwtRequest.OAuth2Extractor}, o.extractors...)
+
+	return func(c *gin.Context) {
+
+		var claims ScopedClaims
+		accessToken, err := jwtRequest.ParseFromRequest(c.Request, extractor, hmacKeyFunc, jwtRequest.WithClaims(&claims))
+
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusUnauthorized,
+				gin.H{
+					"message": "Invalid token",
+					"error":   err.Error(),
+				})
+			return
+		}
+
+		isTokenWhitelisted, err := isTokenWhitelisted(o.tokenStore, &claims)
+
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusInternalServerError,
+				gin.H{
+					"message": "Failed to validate user token",
+					"error":   err.Error(),
+				})
+			log.Println("Failed to lookup user token:", err)
```

Comment:
This should either be removed, or replaced with an injected logger (can fall back to a noop logger) or even better: passed to an error handler. But it could be part of a subsequent PR.

Reply:
Yes, I totally agree.